enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,535 @@
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata: {
+  name: foo
+  annotations: {
+    "olm.Condition.Upgradeable" : "Migrating"
+    "olm.Condition.Important" : "BadConnectivity,UnhealthyDatabase"
+  }
+}
+```
+
+When this annotation is present on a CRD, OLM will append the following `Probe` CR to the `installPlan`:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  specOverride: crdAnnotations # Changes to the CRD's annotations should override the spec. SPO will watch the CRD for changes to the "OLM Supported Condition" annotations and update the spec accordingly.
+  probeResources:
+  - name: foo.example.com 
+    upgradeable: # Converted to array
+    - Migrating
+    important: # Converted to array
+    - BadConnectivity
+    - UnhealthyDatabase
+```
+
+When the`probe` resource above is created, OLM will spin up an SPO pod which collects the information from conditions of `foo.example.com` CRs and communicate it to OLM via the `probe`'s status accordingly. The permissions of the probe will match the same permissions of the operator.
+
+>Note: If the probe's spec is nil, SPO will not create a pod and the operator will be expected to update the CRs status itself.
+
+In this example, the pod will be responsible for maintaining the following fields in the status of the `probe` resource:
+
+```yaml
+status:
+  upgradeable: # A boolean that represents if the operator may be upgraded - aggregated from all probeResources below.
+  probeResources: # The list of resources the probe will watch and the conditions of those resources.
+```
+
+Consider an upgrade where the a single `Foo` CR exists on cluster:
+
+```yaml
+apiVersion: foo.example.com/v1
+kind: Foo
+metadata:
+  name: foo-example
+  namespace: default"
+spec:
+  ...
+  ...
+  ...
+status:
+  probeResources:
+  - resource: "foo.example.com"
+    - uid: "xxx"
+      conditions:
+      - type: Migrating # Condition 1
+        status: true
+        reason: "CriticalProcess" 
+        message: "Migrating old database schema to new format"
+      - type: unhealthyDatabase # Condition 2
+        status: true
+        reason: "LivenessProbeFailure"
+        message: "The foo-database failed its liveness probe."
+      - type: badConnectivity # Condition 3
+        status: true
+        reason: "404" 
+        message: "Unable to reach the database"
+```
+
+From this CR, OLM will update the Probe CR like so:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: Probe
+metadata:
+  name: foo-operator.v1.0.0
+  namespace: operators
+spec:
+  probeResources:
+  - name: foo.example.com
+    upgradeable:
```

Comment:
Won't it be important to know polarity information here?

In this example, `Migrating` is treated as `Upgradeable`. But when `Migrating=true`, I expect `Upgrading=false`

You could also imagine conditions with the opposite polarity, i.e. `NotBusyAtAll: true` should be translated to `Upgradeable: true`

```yaml
 probeResources:
  - resource: foo.example.com
     type: Upgradeable
     name: Migrating
     polarity: AbnormalTrue | NormalTrue
```

or would it be more reasonable to allow boolean expressions over condition types?
```yaml
 probeResources:
  - name: foo.example.com
     upgradeable: !Migrating
```

```yaml
 probeResources:
  - name: foo.example.com
     upgradeable: !Migrating && ReadyToGo
```

Reply:
Agree, in general it seems off that the status of the `upgradeable` condition is not a boolean but could be anything inbetween `migrating`, `badconnectivity`, `unhealthydatabase`. Not all these states would prohibit upgrades. A simple `true` or `false` would go a long way in making this more approachable.

Reply:
I will update the enhancement to include boolean expressions over condition types (option 2).

Reply:
Addressed in the latest version of the PR.